network/src/service/discovery_service.rs

```
@@ -0,0 +1,291 @@
+    event_sender: mpsc::UnboundedSender<DiscoveryEvent>,
+}
+
+impl DiscoveryProtocol {
+    pub fn new(
+        id: ProtocolId,
+        event_sender: mpsc::UnboundedSender<DiscoveryEvent>,
+    ) -> DiscoveryProtocol {
+        let addr_mgr = DiscoveryAddressManager {
+            event_sender: event_sender.clone(),
+        };
+        let discovery = Discovery::new(addr_mgr);
+        let discovery_handle = discovery.handle();
+        DiscoveryProtocol {
+            id,
+            discovery: Some(discovery),
+            discovery_handle,
+            discovery_senders: FnvHashMap::default(),
+            event_sender,
+        }
+    }
+}
+
+impl ServiceProtocol for DiscoveryProtocol {
+    fn init(&mut self, control: &mut ServiceContext) {
+        debug!("protocol [discovery({})]: init", self.id);
+
+        let discovery_task = self
+            .discovery
+            .take()
+            .map(|discovery| {
+                debug!("Start discovery future_task");
+                discovery
+                    .for_each(|()| Ok(()))
+                    .map_err(|err| {
+                        warn!("discovery stream error: {:?}", err);
+                    })
+                    .then(|_| {
+                        debug!("End of discovery");
+                        Ok(())
+                    })
+            })
+            .unwrap();
+        control.future_task(discovery_task);
+    }
+
+    fn connected(&mut self, control: &mut ServiceContext, session: &SessionContext, _: &str) {
+        debug!(
+            "protocol [discovery] open on session [{}], address: [{}], type: [{:?}]",
+            session.id, session.address, session.ty
+        );
+        let event = DiscoveryEvent::Connected {
+            session_id: session.id,
+            peer_id: session.remote_pubkey.clone().map(|pubkey| pubkey.peer_id()),
+        };
+        // TODO: handle send failed
+        self.event_sender
```

ignore the error, the channel may be dropped when progress shutdown.